src/perf/index.ts

```diff
@@ -0,0 +1,133 @@
+import { logger } from '@libp2p/logger'
+import type { IncomingStreamData, Registrar } from '@libp2p/interface-registrar'
+import type { PeerId } from '@libp2p/interface-peer-id'
+import type { Startable } from '@libp2p/interfaces/startable'
+import type { AbortOptions } from '@libp2p/interfaces'
+import type { ConnectionManager } from '@libp2p/interface-connection-manager'
+
+export const PROTOCOL = '/perf/1.0.0'
+
+const log = logger('libp2p:perf')
+
+const writeBlockSize = BigInt(64 << 10)
+const maxStreams = 1 << 10
+
+export interface PerfComponents {
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class PerfService implements Startable {
+  public readonly protocol: string
+  private readonly components: PerfComponents
+  private started: boolean
+  private readonly databuf: ArrayBuffer
+
+  constructor (components: PerfComponents) {
+    this.components = components
+    this.started = false
+    this.protocol = PROTOCOL
+    this.databuf = new ArrayBuffer(Number(writeBlockSize))
+  }
+
+  async start () {
+    await this.components.registrar.handle(this.protocol, (data: IncomingStreamData) => { void this.handleMessage(data) }, {
+      maxInboundStreams: maxStreams,
+      maxOutboundStreams: maxStreams
+    })
+    this.started = true
+  }
+
+  async stop () {
+    await this.components.registrar.unhandle(this.protocol)
+    this.started = false
+  }
+
+  isStarted () {
+    return this.started
+  }
+
+  async handleMessage (data: IncomingStreamData) {
+    const { stream } = data
+
+    let bytesToSendBack: bigint | null = null
+    for await (const buf of stream.source) {
+      if (bytesToSendBack === null) {
+        bytesToSendBack = BigInt(buf.getBigUint64(0, false))
+      }
+      // Ingest all the bufs and wait for the read side to close
+    }
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    if (bytesToSendBack === null) {
+      throw new Error('bytesToSendBack was null')
+    }
+    await stream.sink(async function * () {
+      while (bytesToSendBack > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > bytesToSendBack) {
+          toSend = bytesToSendBack
+        }
+        bytesToSendBack = bytesToSendBack - toSend
+        yield uint8Buf.subarray(0, Number(toSend))
+      }
+    }())
+  }
+
+  async startPerfOnStream (peer: PeerId, sendBytes: bigint, recvBytes: bigint, options: AbortOptions = {}): Promise<void> {
+    log('dialing %s to %p', this.protocol, peer)
```

Comment:
Might include other logs:
Log for finishing a perf
Log for handling/finishing an incoming perf